### PR TITLE
:sparkles: feat(api): Introduce SlimefunItemRegistryFinalizedEvent

### PR DESCRIPTION
**Pull request by ProfElements**
## Description

This Event is provided so that plugins that have integrations with each other aren't stuck in a state in which some SlimefunItems may not be available to one plugin or the other.

I will use this event to provide a cleaner integration between Dynatech and other plugins. Currently InfinityXpansion, ExoticGarden, and Gastronomicon.

## Proposed changes

Make an event to be able to listen to Slimefun's Registry finalizing all static items. 

## Checklist

- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.20.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [x] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Comment by J3fftw1** [comment]
Can we get unit tests?

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunRegistryFinalizedEvent.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItem;
+import io.github.thebusybiscuit.slimefun4.implementation.Slimefun;
+
+
+/**
+ * This {@link Event} is fired after {@link Slimefun} finishes loading the
+ * {@link SlimefunItem} registry before auto loading
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunRegistryFinalizedEvent.java:14]: "before auto loading"
before auto loading what?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/PostSetup.java [hunk 2]
```diff
@@ -109,6 +110,7 @@ public static void loadItems() {
 
         Slimefun.getItemCfg().save();
         Slimefun.getResearchCfg().save();
+        Bukkit.getPluginManager().callEvent(new SlimefunRegistryFinalizedEvent());
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/PostSetup.java:113]: Seems like a bit of a strange place to put this call, any reason it's not higher up? Either right under item load or right under the recipe load.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunRegistryFinalizedEvent.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * @author ProfElements
+ */
+public class SlimefunRegistryFinalizedEvent extends Event {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunRegistryFinalizedEvent.java:18]:
```suggestion
public class SlimefunItemRegistryFinalizedEvent extends Event {
```

**Comment by ProfElements** [comment]
Fixed up according to suggestions

**Comment by JustAHuman-xD** [comment]
I'm not sure about the use case of this tbch, for example: if one addon waits for this event to then register its items to integrate with another addon, but then *another* addon waits for this event to integrate with that addon, then you would still have the same problem.

Currently if you soft depend on any addon you can utilize its items without any problem, in my experience anyways. I guess it depends on the addon but I've never experienced that not working.

The use case you have also provided encourages registering items, after load items is called, which feels wrong. 

Idk this could just be me but I don't think this is the right solution.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/PostSetup.java [hunk 4]
```diff
@@ -109,6 +112,7 @@ public static void loadItems() {
 
         Slimefun.getItemCfg().save();
         Slimefun.getResearchCfg().save();
+
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/PostSetup.java:115]:
```suggestion
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    public static void load() {
+        server = MockBukkit.mock();
+        plugin = MockBukkit.load(Slimefun.class);
+    }
+
+    @AfterAll
+    public static void unload() {
+        MockBukkit.unmock();
+    }
+
```
Comment (JustAHuman-xD) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java:30]:
```suggestion
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Test
+    @DisplayName("Test that SlimefunRegistryFinalizedEvent is fired")
+    void testEventIsFired() {
+
```
Comment (JustAHuman-xD) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java:35]:
```suggestion
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package io.github.thebusybiscuit.slimefun4.api.events;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
```
Comment (JustAHuman-xD) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java:3]: Static imports are against the code style guidelines ([See here](https://github.com/Slimefun/Slimefun4/blob/master/CONTRIBUTING.md#1-imports))

**Review by JustAHuman-xD — CHANGES_REQUESTED** [review]

**Comment by ProfElements** [comment]
> I'm not sure about the use case of this tbch, for example: if one addon waits for this event to then register its items to integrate with another addon, but then _another_ addon waits for this event to integrate with that addon, then you would still have the same problem.
> 
> Currently if you soft depend on any addon you can utilize its items without any problem, in my experience anyways. I guess it depends on the addon but I've never experienced that not working.
> 
> The use case you have also provided encourages registering items, after load items is called, which feels wrong.
> 
> Idk this could just be me but I don't think this is the right solution.

This probably isn't the right solution, but I'm currently in the situation when I want to integrate items into another project. Which I can't load after because its `softdepend` on DynaTech. I don't thing we should add items to registry after this point, I'm using it to add all my recipes using another projects items IE my SeedPlucker adding Gastro Seeds to its recipes

**Comment by JustAHuman-xD** [comment]
> This probably isn't the right solution, but I'm currently in the situation when I want to integrate items into another project. Which I can't load after because its `softdepend` on DynaTech. I don't thing we should add items to registry after this point, I'm using it to add all my recipes using another projects items IE my SeedPlucker adding Gastro Seeds to its recipes


Ah so you're in a circular dependency problem. Which is similar to the issue I showed with this. One solution I can think of is you PR to gastro and make gastro add it to the seed plucker but yeah I get the problem.

I think what you may be able to do is something kinda like slimefun, slimefun adds a lot of recipes **to machines** on post register. So if all you want to do is add machine recipes, then this event is actually a good solution. In which case I misunderstood what this was meant for. I thought you meant registering items dependent on addons items 😅

**Comment by ProfElements** [comment]
Yeah Im sorry i know my explanation was kinda confusing

**Review by JustAHuman-xD — DISMISSED** [review]
LGTM, apologizes for the confusion 😅

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    public static void load() {
+        server = MockBukkit.mock();
+        plugin = MockBukkit.load(Slimefun.class);
+    }
+
+    @AfterAll
+    public static void unload() {
+        MockBukkit.unmock();
+    }
+    @Test
```
Comment (JustAHuman-xD) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java:30]:
```suggestion

    @Test
```

**Review by JustAHuman-xD — COMMENTED** [review]
(You double removed the empty line here, not a big deal I'll re add it lol)

**Review by JustAHuman-xD — APPROVED** [review]
LGTM

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunItemRegistryFinalizedEvent.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+    @Nonnull
+    @Override
+    public HandlerList getHandlers() {
+        return getHandlerList(); 
+    }
+
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunItemRegistryFinalizedEvent.java:34]:
```suggestion
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunItemRegistryFinalizedEvent.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItem;
+import io.github.thebusybiscuit.slimefun4.implementation.Slimefun;
+
+
+/**
+ * This {@link Event} is fired after {@link Slimefun} finishes loading the
+ * {@link SlimefunItem} registry.
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunItemRegistryFinalizedEvent.java:14]:
```suggestion
 * {@link SlimefunItem} registry. We recommend listening to this event if you
 * want to register recipes using items from other addons.
```

**Review by J3fftw1 — DISMISSED** [review]

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    @AfterAll
+    public static void unload() {
+        MockBukkit.unmock();
+    }
+
+    @Test
+    @DisplayName("Test that SlimefunRegistryFinalizedEvent is fired")
+    void testEventIsFired() {
+        //Make sure post setup does not throw
```
Comment (WalshyDev) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java:34]:
```suggestion
        // Make sure post setup does not throw
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        Assertions.assertDoesNotThrow(() -> PostSetup.loadItems());
+        
+        //Make sure post setup sent the event
```
Comment (WalshyDev) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java:37]:
```suggestion
        // Make sure post setup sent the event
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        server.getPluginManager().assertEventFired(SlimefunItemRegistryFinalizedEvent.class, ignored -> true);
+ 
+        server.getPluginManager().clearEvents();
+    }
+
```
Comment (WalshyDev) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestSlimefunRegistryFinalizedEvent.java:42]:
```suggestion
```

**Review by WalshyDev — APPROVED** [review]

**Review by J3fftw1 — APPROVED** [review]

**Pull request merged**